asm/exceptasm.s

```diff
@@ -69,646 +81,891 @@ glabel __osIntTable
+    mtc0    $t0, C0_CAUSE
+    # Post sw2 event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_SW2*8
+    li      $at, ~CAUSE_SW2
+    # Mask out interrupt and continue
+    b       next_interrupt
+     and    $s0, $s0, $at
+
 sw1:
-/* 004A78 80003E78 2401FEFF */  li    $at, -257
-/* 004A7C 80003E7C 01014024 */  and   $t0, $t0, $at
-/* 004A80 80003E80 40886800 */  mtc0  $t0, $13
-/* 004A84 80003E84 0C000FCB */  jal   send_mesg
-/* 004A88 80003E88 24040000 */   li    $a0, 0
-/* 004A8C 80003E8C 2401FEFF */  li    $at, -257
-/* 004A90 80003E90 1000FF53 */  b     next_interrupt
-/* 004A94 80003E94 02018024 */   and   $s0, $s0, $at
+    li      $at, ~CAUSE_SW1
+    # Clear interrupt
+    and     $t0, $t0, $at
+    mtc0    $t0, C0_CAUSE
+    # Post sw1 event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_SW1*8
+    li      $at, ~CAUSE_SW1
+    # Mask out interrupt and continue
+    b       next_interrupt
+     and    $s0, $s0, $at
+
 handle_break:
-/* 004A98 80003E98 24090001 */  li    $t1, 1
-/* 004A9C 80003E9C A7490012 */  sh    $t1, 0x12($k0)
-/* 004AA0 80003EA0 0C000FCB */  jal   send_mesg
-/* 004AA4 80003EA4 24040050 */   li    $a0, 80
-/* 004AA8 80003EA8 10000001 */  b     redispatch
-/* 004AAC 80003EAC 00000000 */   nop   
+    # Set thread as having hit a break exception
+    li      $t1, OS_FLAG_CPU_BREAK
+    sh      $t1, THREAD_FLAGS($k0)
+    # Post a cpu break event message
+    jal     send_mesg
+     li     $a0, OS_EVENT_CPU_BREAK*8
+    # Redispatch
+    b       redispatch
+     nop
+
 redispatch:
-/* 004AB0 80003EB0 3C0A8001 */  lui   $t2, %hi(__osRunQueue) # $t2, 0x8001
-/* 004AB4 80003EB4 8D4AAD48 */  lw    $t2, %lo(__osRunQueue)($t2)
-/* 004AB8 80003EB8 8F490004 */  lw    $t1, 4($k0)
-/* 004ABC 80003EBC 8D4B0004 */  lw    $t3, 4($t2)
-/* 004AC0 80003EC0 012B082A */  slt   $at, $t1, $t3
-/* 004AC4 80003EC4 10200007 */  beqz  $at, enqueueRunning
-/* 004AC8 80003EC8 00000000 */   nop   
-/* 004ACC 80003ECC 3C048001 */  lui   $a0, %hi(__osRunQueue) # $a0, 0x8001
-/* 004AD0 80003ED0 03402825 */  move  $a1, $k0
-/* 004AD4 80003ED4 0C001045 */  jal   __osEnqueueThread
-/* 004AD8 80003ED8 2484AD48 */   addiu $a0, %lo(__osRunQueue) # addiu $a0, $a0, -0x52b8
-/* 004ADC 80003EDC 0800105D */  j     __osDispatchThread
-/* 004AE0 80003EE0 00000000 */   nop   
+    lui     $t2, %hi(__osRunQueue)
+    lw      $t2, %lo(__osRunQueue)($t2)
+    # Get currently running priority
+    lw      $t1, THREAD_PRI($k0)
+    # Get highest priority from waiting threads
+    lw      $t3, THREAD_PRI($t2)
+    slt     $at, $t1, $t3
+    beqz    $at, enqueueRunning
+     nop
+    # The current thread is no longer the highest priority, enqueue it
+    #  to the run queue to wait its turn again
+    lui     $a0, %hi(__osRunQueue)
+    move    $a1, $k0
+    jal     __osEnqueueThread
+     addiu  $a0, $a0, %lo(__osRunQueue)
+    j       __osDispatchThread
+     nop
 
+/**
+ *  Dispatches the top of the run queue
+ */
 enqueueRunning:
```

Comment:
and if I understand it right it also shifts the "current thread"/saved interrupted thread context (k0) at the top of the run queue before dispatching, so really it's just resuming the thread isn't it?

(to be clear the comment should probably mention how this also changes the run queue and that the dispatched thread is the "current" one)